Sources_v3/StreamChat/ChatClient.swift

```diff
@@ -271,54 +271,195 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
     ///
     init(
         config: ChatClientConfig,
+        tokenProvider: _TokenProvider<ExtraData>,
         workerBuilders: [WorkerBuilder],
         eventWorkerBuilders: [EventWorkerBuilder],
         environment: Environment
     ) {
+        log.assert(Thread.isMainThread, "ChatClient.init must be called on main thread.")
+
         self.config = config
+        self.tokenProvider = tokenProvider
         self.environment = environment
         self.workerBuilders = workerBuilders
         self.eventWorkerBuilders = eventWorkerBuilders
-        
+        currentUserId = databaseContainer.viewContext.currentUser()?.user.id
+
         createBackgroundWorkers()
         createWebSocketClient()
     }
     
     deinit {
-        _connectionIdWaiters.mutate { waiters in
-            waiters.forEach { $0(nil) }
-            waiters.removeAll()
-        }
+        removeConnectionWaitersPassing(nil)
+        removeTokenWaitersPassing(nil)
     }
-    
-    // TODO: Not used & tested yet -> CIS-224
-    func refreshToken(completion: @escaping (Error?) -> Void) {
-        log.assert(config.tokenProvider != nil, "You can't call `refreshToken` when `Config.tokenProvider` is nil.")
-        
-        config.tokenProvider?(config.apiKey, currentUserId, { [weak self] (token) in
-            guard let token = token else {
-                completion(ClientError.MissingToken("Can't connect because `TokenProvider` didn't return a valid token."))
-                return
-            }
-            self?.currentToken = token
-            completion(nil)
-        })
-    }
-    
+}
+
+private extension _ChatClient {
     func createBackgroundWorkers() {
         backgroundWorkers = workerBuilders.map { builder in
             builder(self.databaseContainer, self.apiClient)
         } + eventWorkerBuilders.map { builder in
             builder(self.databaseContainer, self.eventNotificationCenter, self.apiClient)
         }
     }
-    
+
+    func removeConnectionWaitersPassing(_ connectionId: String?) {
```

Comment:
This and its `tokenWaiters` friend's naming confused me, maybe `removeConnectionIdWaiters(passing value: nil)`?

Reply:
+1 on confusing naming. Why not just: `completeConnectionIdWaiters(value: String?)`? It work nicely with both an actual value and `nil`, too.

Reply:
Great idea, will fix it

Reply:
fixed